### PR TITLE
MLP Output Norm Fixes

### PR DESCRIPTION
**Pull request by bkal01**
- log mlp output norm to see how norm explosion is happening
- add configurable RMSNorm after the MLP projector

right now the $$L_2$$-norm of the MLP connector's output is 2-3 orders of magnitude higher than the $$L_2$$-norm of the LLM token embeddings.
Screenshot 2026-03-12 at 11 52 39 PM
Screenshot 2026-03-12 at 11 53 03 PM

there's an RMSNorm before attention and before the MLP in each transformer block, but after attention we add the residual stream which is not normalized. so our pre-MLP RMSNorm looks like $$\text{RMSNorm}(r + a)$$, where $$r$$ is the unnormalized residual and $$a$$ is the output of attention.

$$\text{RMSNorm}(r + a) = \frac{r + a}{||r+a||_2} \cdot \sqrt{d} \approx \frac{r + a}{||r||_2}\cdot\sqrt{d} = (\frac{r}{||r||_2} + \frac{a}{||r||_2})\cdot\sqrt{d} \approx \text{RMSNorm}(r)$$

So the MLP connector output will drown out the output of attention/MLP in the transformer blocks

if we normalize the connector output then this issue should be mitigated, i have an ongoing run with this normalization applied

### CODE REVIEW

File: pipeline/train_alignment.py [hunk 1]
```diff
@@ -67,9 +70,21 @@ def train(
                     pixel_values=pixel_values,
                     labels=labels,
                 )
-                loss = outputs.loss / training_config.grad_acc_steps
+                ce_loss = outputs.loss / training_config.grad_acc_steps
+
+                token_embeddings = model.language_model.get_input_embeddings().weight # (vocab size, D)
```
[the remark this answers is not in the document]
Comment (d-rau) [reply pipeline/train_alignment.py]: you are absolutely right! I missed this
